### PR TITLE
PCA feature

### LINKED ISSUE #168: Add an analysis tool for PCA/dimensionality reduction

### PR DESCRIPTION
**Pull request by Alvaroguleon**
fixes #168
The setup.py has been updated so that it yields no errors when installing MutateX from scratch. I have not included adjustText as a requirement though, so it will probably need to be installed manually.

 he version of ddg2pca committed plots a PCA biplot and a scree plot. The default PCA plot groups the residues of the PDB file according to their nature and shows a scatter and a text visualization. The color and group of the residues can be changed with the configuration file, the textfile can be used to select a subset of label texts to see, and the poslist can be used to select a subset of residues to see overall.

### CODE REVIEW

File: bin/ddg2pca [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+    # Replacing the res_str in the data by the corresponding labels 
+    for i, label in enumerate(filtered_labels):
+        data[i, 0] = label
+
+    return data, column_headers, filtered_labels, filtered_textlist, filtered_residue_grouping, color_dict
+
+    # np.savetxt(f"{options.name_outfile}.csv", csv_table, delimiter=",", fmt='%s')  
+def perform_pca(data):
+    
+    data_scaled = (data[:, 1:].astype(float) - np.mean(data[:, 1:].astype(float), axis=0)) / np.std(data[:, 1:].astype(float), axis=0)
```
Comment (mtiberti) [inline bin/ddg2pca:263]: it would be good to have an option not to rescale data (while I think we should always center them). After all, all our data sets are on the same scale by definition

File: setup.py [hunk 2]
```diff
@@ -35,12 +36,13 @@
               'bin/ddg2summary'],
 
     install_requires=[ 'biopython',
-                       'matplotlib',
-		       'numpy',
-                       'scipy',
+                       'matplotlib==3.4.3',
+		       'numpy==1.21.0',
+                       'scipy==1.6.3',
```
Comment (mtiberti) [inline setup.py:41]: can we do these changes in a separate PR, with its own issue?

**Review by mtiberti — CHANGES_REQUESTED** [review]

File: bin/ddg2pca [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        except (IOError, TypeError):
+            exit(1)
+
+        textlist = filter_reslist(res_ids_str, textlist)
+
+        # Determine if textlist items are single strings or tuples
+        is_single_chain = len(textlist[0]) == 1 # textlist items are single strings or tuples   
+
+    else:
+        textlist = []
```
[the remark this answers is not in the document]
Comment (Alvaroguleon) [reply bin/ddg2pca:140]: There is actually a reason for this. Later in the code, I apply some adjustments for the textlist for 1) Change the names in the textlist according to the new labels if these are provided and 2) Subset the textlist according to the specified residue gropung if provided. I noticed that if I created an empty textlist, the rest of the code could be maintained as is. However, if I set the textlist to None, the steps 1 and 2 returned errors due to texlist being nonexistent, and I would have to implement in each of those steps a condition ("if textlist..") before the code that adjusts the texlist to avoid bugs in case of a None texlist. I thought the empty list was a cleaner way to do it. Based on this, let me know if you'd still prefer the None version and I will make the changes

File: setup.py (hunk 2 above)
Comment (Alvaroguleon) [reply setup.py:41]: Absolutely, I can create the issue myself and I will revert the changes in the setup.py. The only question that I have is whether I should also create a new branch for the change in the setup.py and create the PR from there

File: bin/ddg2pca (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtiberti) [reply bin/ddg2pca:140]: I think the question here is more on what behaviour the user will expect; in this case, an empty file would result in the default without textfile, e.g. all the labels rather than in no labels which I think would be the expected outcome

File: setup.py (hunk 2 above)
Comment (mtiberti) [reply setup.py:41]: yep - new issue, connected new branch _from the `master` branch_ and associated PR

File: bin/ddg2pca (hunk 3 above)
[the remark this answers is not in the document]
Comment (Alvaroguleon) [reply bin/ddg2pca:140]: Hmmm I see your point. That interpretation of the code did not come to my mind. I will change it to None then :)

**Review by mtiberti — APPROVED** [review]
